### PR TITLE
MDR, DD, SR Mandate Submission

### PR DESCRIPTION
**Pull request by lomky**
Bumps [CASEFLOW-98](https://vajira.max.gov/browse/CASEFLOW-98)

### Description
Allows for the submission of the Mandate form on MDR, DD, and SR.

Stack PR for [CASEFLOW-98](https://vajira.max.gov/browse/CASEFLOW-98)

Part 1 - Adds the Remand Received action to MDR Task #15860
Part 2 - Allows for the submission of the Mandate form on MDR, DD, and SR.
Part 3 - TK FeatureTest & seeds

### Acceptance Criteria
 - [ ] Work is behind the feature toggle: mdr_cavc_remand
 - [ ] Accessible to the CAVC Litigation Support team
 - [ ] Verify confirmation banner is presented on the users queue when submit is clicked on the new page with the following copy:
"You have successfully created a CAVC Remand case
This has generated a 90 day letter task in your organizations queue"
 - [ ]  On submission of the form for MDR
   - the missing dates and instructions on an MDR are updated
   - the MDR task is completed
   - the 90 day letter task is created in CAVC Litigation Supports queue
 - [ ]  On submission of the form for SR & DD
   - the missing dates and instructions on an MDR are updated
   - the MDR & CAVC task are completed
   - the Distribution Task is active
   - the success detail says it went to distribution _instead_ of to CAVC Lit Support queue

### Testing Plan
1. Create an MDR cavc remand
    `FactoryBot.create(:cavc_remand, :mdr).remand_appeal.uuid`
2. Sign in as a CAVC Lit Support user and navigate to the appeal
3. fill out the form & submit
   - [ ] Confirm it submits & shows the message
   - [ ] Confirm the appeal has its send letter task assigned to CavcLitSupport
4. Repeat for No mandate Straight Reversal
    `FactoryBot.create(:cavc_remand, :straight_reversal, :no_mandate).remand_appeal.uuid`
   - [ ] Confirm it submits & shows the message
   - [ ] Confirm the appeal has an active Distribution Task
5. Repeat for No mandate Death Dismissal
   `FactoryBot.create(:cavc_remand, :death_dismissal, :no_mandate).remand_appeal.uuid`
   - [ ] Confirm it submits & shows the message
   - [ ] Confirm the appeal has an active Distribution Task

### CODE REVIEW

File: app/models/cavc_remand.rb [hunk 1]
```diff
@@ -52,13 +60,30 @@ def cavc_remand_form_complete?
     valid? && (mandate_not_required? || (!mandate_date.nil? && !judgement_date.nil?))
   end
 
+  def already_has_mandate?
+    !mandate_not_required? || (!remand? && mandate_date)
+  end
+
   def establish_appeal_stream
-    self.remand_appeal = source_appeal.create_stream(:court_remand).tap do |cavc_appeal|
+    self.remand_appeal ||= source_appeal.create_stream(:court_remand).tap do |cavc_appeal|
```
Comment (lomky) [inline app/models/cavc_remand.rb:68]: Prevents creating an additional remand appeal on update! :scream_cat:

File: app/controllers/cavc_remands_controller.rb [hunk 2]
```diff
@@ -30,28 +37,38 @@ class CavcRemandsController < ApplicationController
   ].flatten.freeze
 
   def create
-    cavc_remand = CavcRemand.create!(create_params)
-    cavc_appeal = cavc_remand.remand_appeal.reload
-    render json: { cavc_remand: cavc_remand, cavc_appeal: cavc_appeal }, status: :created
+    new_cavc_remand = CavcRemand.create!(create_params)
+    cavc_appeal = new_cavc_remand.remand_appeal.reload
+    render json: { cavc_remand: new_cavc_remand, cavc_appeal: cavc_appeal }, status: :created
```
Comment (lomky) [inline app/controllers/cavc_remands_controller.rb:42]: namespace cleanup

File: app/models/cavc_remand.rb [hunk 3]
```diff
@@ -52,13 +60,30 @@ def cavc_remand_form_complete?
     valid? && (mandate_not_required? || (!mandate_date.nil? && !judgement_date.nil?))
   end
 
+  def already_has_mandate?
+    !mandate_not_required? || (!remand? && mandate_date)
+  end
```
Comment (lomky) [inline app/models/cavc_remand.rb:65]: We don't support updates beyond receiving the mandate!

File: client/app/queue/AddCavcDatesModal.jsx [hunk 4]
```diff
@@ -49,7 +50,7 @@ const AddCavcDatesModal = ({ appealId, error, highlightInvalid, history }) => {
       detail: COPY.CAVC_REMAND_CREATED_DETAIL
     };
 
-    requestPatch(`/appeals/${appealId}/cavc_remand`, payload, successMsg).
+    dispatch(requestPatch(`/appeals/${appealId}/cavc_remand`, payload, successMsg)).
```
Comment (lomky) [inline client/app/queue/AddCavcDatesModal.jsx:53]: Tweak to the FE code to tie it together, now that we have a real backend to test with!

File: spec/controllers/cavc_remands_controller_spec.rb [hunk 5]
```diff
@@ -119,18 +132,81 @@
         let(:remand_subtype) { nil }
 
         include_examples "works without judgement and mandate date parameters"
-      end
+     end
     end
 
-    context "without a Lit Support User" do
```
Comment (lomky) [inline spec/controllers/cavc_remands_controller_spec.rb:125]: Moved to shared example

File: client/app/queue/AddCavcDatesModal.jsx [hunk 6]
```diff
@@ -44,12 +49,21 @@ const AddCavcDatesModal = ({ appealId, error, highlightInvalid, history }) => {
...
     };
 
+    const successMsgDetail = () => {
+      if (straightReversalType() || deathDismissalType()) {
+        return COPY.CAVC_REMAND_READY_FOR_DISTRIBUTION_DETAIL;
+      }
+      else {
+        return COPY.CAVC_REMAND_CREATED_DETAIL
+      }
+    };
```
Comment (lomky) [inline client/app/queue/AddCavcDatesModal.jsx:59]: Give the right message on submit

**Review by n1zyy — APPROVED** [review]
👍

File: spec/models/cavc_remand_spec.rb [hunk 7]
```diff
@@ -165,4 +165,93 @@
...
+    end
+
+    subject { cavc_remand.update(params) }
+
+    context "on a JMR appeal" do
+      let(:cavc_remand) { create(:cavc_remand) }
+      it "throws an error" do
+        expect { subject }.to raise_error(Caseflow::Error::MandatedRemandsNoUpdate)
+      end
+    end
```
[the remark this answers is not in the document]
Comment (lomky) [reply spec/models/cavc_remand_spec.rb:190]: I feel like the context the line above provides that?

[the remark this answers is not in the document]
Comment (n1zyy) [reply spec/models/cavc_remand_spec.rb:190]: Ha, that's very fair. 👍

**Pull request merged**